onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -732,7 +732,8 @@ class PlannerImpl {
           AllocPlan(current).alloc_kind = AllocKind::kAllocate;
           AllocPlan(current).program_counter.AddStart(program_counter);
         } else if (!context_.IsParallelExecutionEnabled() &&
-                   FindReusableInput(*pnode, static_cast<int>(output_arg_def_index), &reused)) {
+                   FindReusableInput(*pnode, static_cast<int>(output_arg_def_index), &reused) &&
+                   (pnode->OutputNodesBegin() == pnode->OutputNodesEnd() || pnode->OutputNodesBegin()->OpType() != "YieldOp")) {
           // Reuse one of this node's input buffers as the output buffer (for in-place update)
```

Comment:
Why is this change necessary? What is the issue if the input to YieldOp is from a re-used buffer? 

Would it be better to do this check in FindReusableInput?

Would be good if somewhat obtuse condititions like this have a comment in the code explaining them. 

Is the `pnode->OutputNodesBegin() == pnode->OutputNodesEnd()` necessary? If the node has no output edges would it be producing a graph output and have gone into the 'else if' on line 691?

Reply:
Given we tag the YieldOp kernel with KernelDef::HasExternalOutputs(), could we instead make that tag more generic and also special case the inputs based on it? That way we don't need a hardcoded check on OpType being "YieldOp".

Reply:
The reason for this change is added to the Description of this PR. 

Good point merging this into FindReusableInput(), done. 

HasExternalOutputs() flag is to indicate YieldOp's **output** buffer is allocated by pytorch, and owner by pytorch, and other tensor cannot reuse.

Reply:
I am still keeping pnode->OutputNodesBegin() == pnode->OutputNodesEnd() check, as a defensive programming practice.

Reply:
> HasExternalOutputs() flag is to indicate YieldOp's **output** buffer is allocated by pytorch, and owner by pytorch, and other tensor cannot reuse. This change for the **input** of YieldOp.

I realize that - it was more a question whether there is any independent scenario where we'd have externally allocated outputs, or whether this special casing is really around a node that's an integration point to an external system (and due to that both inputs and outputs need special casing). I was imagining that HasExternalOutputs would become something like IsExternaIntegrationPoint().